File: java/cours/src/session3/TD/MotIterableAvecIterateurEtVisiteur/MotConcat.java
package session3.TD.MotIterableAvecIterateurEtVisiteur;

class MotConcat implements Mot {
	private Mot gauche;
	private Mot droit;
	private int taille;

	public MotConcat(Mot g, Mot d) {
		this.gauche = g;
		this.droit = d;
		this.taille = g.taille() + d.taille();
	}

	@Override
	public int taille() {
		return this.taille;
	}

	@Override
	public boolean casUnion() {
		return true;
	}

	@Override
	public char element() {
		Iterateur i = this.iterateur();
		return i.suivant();
	}

	@Override
	public Mot reste() {
		Iterateur i = this.iterateur();
		return i.reste();
	}

	@Override
	public Mot gauche() {
		return this.gauche;
	}

	@Override
	public Mot droit() {
		return this.droit;
	}
}

File: typescript/session3/MotIterableAvecIterateurEtVisiteur/MotConcat.ts
import { Mot } from "./Mot";
import { Char } from "../Char";
import { Iterateur } from "./Iterateur";

export class MotConcat extends Mot {

  private motGauche: Mot;
  private motDroit: Mot;
  private tailleMot: number;

  public constructor(g: Mot, d: Mot) {
    super();
    this.motGauche = g;
    this.motDroit = d;
    this.tailleMot = g.taille() + d.taille();
  }

  public taille(): number {
    return this.tailleMot;
  }

  public casUnion(): boolean {
    return true;
  }

  public element(): Char {
    let i: Iterateur = this.iterateur();
    return i.suivant();
  }
  public reste(): Mot {
    let i: Iterateur = this.iterateur();
    return i.reste();
  }

  public gauche(): Mot {
    return this.motGauche;
  }
  public droit(): Mot {
    return this.motDroit;
  }
}